### PR TITLE
Ignore expired tasks & clean task executor threads after a max duration

### PR DESCRIPTION
**Pull request by zguesmi**

### CODE REVIEW

File: src/main/java/com/iexec/core/task/Task.java [hunk 1]
```diff
@@ -47,11 +48,13 @@
     @Version
     private Long version;
 
+    @Indexed(unique = true)
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/task/Task.java:51]: Thanks!

**Comment by zguesmi** [comment]
#384

File: src/main/java/com/iexec/core/chain/DealWatcherService.java [hunk 2]
```diff
@@ -54,69 +55,106 @@ public DealWatcherService(IexecHubService iexecHubService,
...
-                        chainDeal.getTag());
-                optional.ifPresent(task -> applicationEventPublisher.publishEvent(new TaskCreatedEvent(task.getChainTaskId())));
-            }
-        } catch (Exception e) {
-            e.printStackTrace();
+        ChainDeal chainDeal = oChainDeal.get();
+        // do not process deals after deadline
+        if (!iexecHubService.isBeforeContributionDeadline(chainDeal)) {
+            log.error("Deal has expired [chainDealId:{}, deadline:{}]",
+                    chainDealId, iexecHubService.getChainDealContributionDeadline(chainDeal));
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/chain/DealWatcherService.java:111]: Maybe in the future we could have at least 2 iexechub services, one for mutable objects and one for immutable objects (here I see we properly save blockchain calls but iexecHubService.getChainDealContributionDeadline(..) lets think the opposite)

File: src/main/java/com/iexec/core/task/TaskService.java [hunk 3]
```diff
@@ -63,16 +63,45 @@ public TaskService(TaskRepository taskRepository,
...
+            String imageName,
+            String commandLine,
+            int trust,
+            long maxExecutionTime,
+            String tag
+    ) {
+        boolean isPresent = taskRepository
+                .findByChainDealIdAndTaskIndex(chainDealId, taskIndex)
+                .isPresent();
+        if (isPresent) {
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/task/TaskService.java:92]: Maybe we could have a full functional-like block here?

**Review by jeremyjams — APPROVED** [review]
CustomSingleExecutorThreadFactory :clap: :pray:

File: src/main/java/com/iexec/core/chain/DealWatcherService.java (hunk 2 above)
Comment (zguesmi) [reply src/main/java/com/iexec/core/chain/DealWatcherService.java:111]: Yes, we should get the most out of immutable objects.

File: src/main/java/com/iexec/core/task/TaskService.java (hunk 3 above)
Comment (zguesmi) [reply src/main/java/com/iexec/core/task/TaskService.java:92]: Done, much cleaner thanks!

**Pull request merged**